src/nnet3/discriminative-supervision.cc

```diff
@@ -0,0 +1,450 @@
+  std::swap(weight, other->weight);
+  std::swap(num_sequences, other->num_sequences);
+  std::swap(frames_per_sequence, other->frames_per_sequence);
+  std::swap(num_ali, other->num_ali);
+  std::swap(den_lat, other->den_lat);
+}
+
+bool DiscriminativeSupervision::operator == (
+    const DiscriminativeSupervision &other) const {
+  return ( weight == other.weight && 
+      num_sequences == other.num_sequences &&
+      frames_per_sequence == other.frames_per_sequence &&
+      num_ali == other.num_ali &&
+      fst::Equal(den_lat, other.den_lat) );
+}
+
+void DiscriminativeSupervision::Write(std::ostream &os, bool binary) const {
+  WriteToken(os, binary, "<DiscriminativeSupervision>");
+  WriteToken(os, binary, "<Weight>");
+  WriteBasicType(os, binary, weight);
+  WriteToken(os, binary, "<NumSequences>");
+  WriteBasicType(os, binary, num_sequences);
+  WriteToken(os, binary, "<FramesPerSeq>");
+  WriteBasicType(os, binary, frames_per_sequence);
+  KALDI_ASSERT(frames_per_sequence > 0 &&
+               num_sequences > 0);
+  
+  WriteToken(os, binary, "<NumAli>");
+  WriteIntegerVector(os, binary, num_ali);
+
+  WriteToken(os, binary, "<DenLat>");
+  if (!WriteLattice(os, binary, den_lat)) {
+    // We can't return error status from this function so we
+    // throw an exception. 
+    KALDI_ERR << "Error writing denominator lattice to stream";
+  }
+
+  WriteToken(os, binary, "</DiscriminativeSupervision>");
+}
+
+void DiscriminativeSupervision::Read(std::istream &is, bool binary) {
+  ExpectToken(is, binary, "<DiscriminativeSupervision>");
+  ExpectToken(is, binary, "<Weight>");
+  ReadBasicType(is, binary, &weight);
+  ExpectToken(is, binary, "<NumSequences>");
+  ReadBasicType(is, binary, &num_sequences);
+  ExpectToken(is, binary, "<FramesPerSeq>");
+  ReadBasicType(is, binary, &frames_per_sequence);
+  KALDI_ASSERT(frames_per_sequence > 0 && 
+               num_sequences > 0);
+  
+  ExpectToken(is, binary, "<NumAli>");
+  ReadIntegerVector(is, binary, &num_ali);
+
+  ExpectToken(is, binary, "<DenLat>");
+  {
+    Lattice *lat = NULL;
+    if (!ReadLattice(is, binary, &lat) || lat == NULL) {
+      // We can't return error status from this function so we
+      // throw an exception. 
+      KALDI_ERR << "Error reading Lattice from stream";
+    }
+    den_lat = *lat;
+    delete lat;
+    TopSort(&den_lat);
+  }
+
+  ExpectToken(is, binary, "</DiscriminativeSupervision>");
+}
+
+bool LatticeToDiscriminativeSupervision(const std::vector<int32> &num_ali,
+                                        const Lattice &den_lat, 
+                                        BaseFloat weight,
+                                        DiscriminativeSupervision *supervision) {
+  supervision->weight = weight;
+  supervision->num_sequences = 1;
+  supervision->frames_per_sequence = num_ali.size();
+  supervision->num_ali = num_ali;
+  supervision->den_lat = den_lat;
+  TopSort(&(supervision->den_lat));
+
+  supervision->Check();
+
+  return true;
```

Comment:
It would never fail- would indicate a serious bug in the lattice.
I think it should be a class member anyway.  Initialize() function would
make more sense.


On Mon, Mar 7, 2016 at 6:45 PM, Vimal Manohar <notifications@github.com>
wrote:

> In src/nnet3/discriminative-supervision.cc
> <https://github.com/kaldi-asr/kaldi/pull/563#discussion_r55294073>:
>
> > +}
> > +
> > +bool LatticeToDiscriminativeSupervision(const std::vector<int32> &num_ali,
> > +                                        const Lattice &den_lat,
> > +                                        BaseFloat weight,
> > +                                        DiscriminativeSupervision *supervision) {
> > +  supervision->weight = weight;
> > +  supervision->num_sequences = 1;
> > +  supervision->frames_per_sequence = num_ali.size();
> > +  supervision->num_ali = num_ali;
> > +  supervision->den_lat = den_lat;
> > +  TopSort(&(supervision->den_lat));
> > +
> > +  supervision->Check();
> > +
> > +  return true;
>
> I think I intended to return false if TopSort fails. I can check that and
> return false if it fails.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/563/files#r55294073>.
>